Zeroconf communication and encryption

vit3k opened the pull request:
Encryption implementation based on librespot to bypass failing SpZeroConfGetVars.
Changed packaging to external libs rather jar with dependencies to eliminate libs extraction and repackaging.
Set java 8.

Unfortunately I can't get jmdns library to work so as a workaround i've just used avahi-publish tools. It's available in avahi-utils package

vit3k committed:
Zeroconf communication and encryption

Encryption implementation based on librespot to bypass failing SpZeroConfGetVars.
Changed packaging to external libs rather jar with dependencies to eliminate libs extraction and repackaging.
Set java 8.

Unfortunately I can't get jmdns library to work so as a workaround i've just used avahi-publish tools - for ubuntu avahi-utils.

LeanderK commented:
shouldn't that be optional? I mean great work, but i would like to have the things replacing the library optional, so you could still use this as a minimal wrapper.

crsmoro commented:
- Encryption implementation based on librespot to bypass failing SpZeroConfGetVars. - Great!
- Changed packaging to external libs rather jar with dependencies to eliminate libs extraction and repackaging. - Disagree, the idea is exactly to create the jar with dependencies to copy/paste anywhere and just run without any additional worries.
- Set java 8. - No problem.

About the jmdns/avahi-publish: on openelec where i run scplayer today, it give's me an not found error because openelec doesn't have 'ahavi-publish', the ideal would be get jmdns to work. for testing purposes today or tomorrow i will install another that allow me to install avahi-utils..

vit3k committed:
Zeroconf communication and encryption

Code refactored.
Reverted changes to packaging.

vit3k commented:
I've reverted change to packaging. I've changed that in the first place because I wanted to use BouncyCastle and assembly plugin "unzip" each dependency jar and merge it with app's code. This causes a problem with signed jars because they're not a jar anymore and JCE won't add crypto provider which doesn't come from signed jar. But BouncyCastle is not used anymore so it is not needed.

I've refactored a code to make it more object oriented, make "opensource" zeroconf implementation optional by zeroconf system property (library - use spotify library - not implemented as I can't get it to work properly, opensource - my custom implementation).

For now I can't get jmdns to work so for now avahi-publish is the easiest option. I use osmc which is full os so it's more flexible that's why I like it over openelec.

vit3k committed:
Zeroconf communication and encryption

crsmoro commented:
i installed osmc, there was no errors, but nothing happened.. do i have to do something in peculiar?

i prefer openelec over osmc for the exactly same reason..
i have another pi to handle all the over stuff, the one with scplayer/kodi it's only for media center.. keep the pi way faster..

vit3k commented:
Firstly after last update you need to specify

```
 -Dzeroconf=opensource 
```

to use my custom implementation. By default is used library implementation which in fact is not implemented.
Also nothing really should happen on server side. SCplayer should show up on every client in local network. You could also check if service is advertised by running:

```
avahi-browse --all 
```

to list services. You should see one with  _spotify-connect._tcp type.

crsmoro commented:
sorry, but this is not working.. i was forgetting `-Dzeroconf=opensource` ..
now it shows in `avahi-browse --all`
but even with that it doesnt authenticate scplayer when i open the client on windows or iphone..
it supposed to do that, right?

vit3k commented:
I've tested it with android devices and I don't have any ios right now. I will check on Windows later. Can you post log (with -Ddebug=3 option)? Do you see any tries to access getInfo in log?

crsmoro commented:
there's no much log..
it tries to access getInfo once, and then nothing..
nothing shows on spotify clients..
but it is still running cuz if i open the browser and login manually everything is normal..

vit3k commented:
For me it works on all my android devices, windows 10, osx but it's not working on windows 8.1 - I'm not sure why. Also I don't have ios device to try.

It looks like on windows 8.1 zeroconf is not working at all as nothing shows up in scplayer's logs. Scplayer only shows up if logged in account.

What's more it looks like on windows and android sometimes there is need to restart the client. On osx it shows up almost immediately.

I'm not sure how to resolve those issues.

vit3k committed:
zerconf - native lib implementation, lib headers updated

vit3k commented:
I've added lib's native implementation. JNA headers were updated. Current version is working with following library:
https://github.com/Fornoth/spotify-connect-web/files/85265/libspotify-embedded-shared.zip

Original post: https://github.com/Fornoth/spotify-connect-web/issues/25

Fornoth commented:
I put both the armel (soft float) and hard float binaries here in a more permanent location https://github.com/sashahilton00/spotify-connect-resources/tree/master/libs and a updated header file is here https://github.com/Fornoth/spotify-connect-web/blob/master/spotify.h

crsmoro commented:
sorry taking so long, i was out of town.. now it's working!
i will update the links to the libs on readme too..
